### PR TITLE
Refactor, DRY AddPaymentMethodViewController

### PR DESCRIPTION
**Pull request by yuki-stripe**
## Summary
`AddPaymentMethodViewController` (APMVC) originally controlled both the form *and* the horizontal list of PMs, all in one. 

This PR **moves** form-related logic from APMVC to `PaymentMethodFormViewController` (PMFVC), so that it can be reused by vertical mode.  It also makes APMVC **use* PMFVC.

Some big bits of logic that was moved:
- Creating paymentOption from form. Link and Instant Debits both have special hacks.
- Default billing <> shipping address code in viewWillAppear

## Motivation
https://jira.corp.stripe.com/browse/MOBILESDK-2130

## Testing

- Form cache is reset on load
- US Bank and others preserve details when switching between forms
- Shipping details updates form

## Changelog
Not user facing

### CODE REVIEW

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/New Payment Method Screen/AddPaymentMethodViewController.swift [hunk 1]
```diff
@@ -134,44 +107,27 @@ class AddPaymentMethodViewController: UIViewController {
 
     private let intent: Intent
     private let configuration: PaymentSheet.Configuration
+    private let isLinkEnabled: Bool
```
Comment (porter-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/New Payment Method Screen/AddPaymentMethodViewController.swift:110]: Maybe naive: But can we read is link enabled from the intent?

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/ViewControllers/PaymentMethodFormViewController.swift [hunk 2]
```diff
@@ -5,40 +5,142 @@
...
 
+protocol PaymentMethodFormViewControllerDelegate: AnyObject {
+    func didUpdate(_ viewController: PaymentMethodFormViewController)
+    func updateErrorLabel(for error: Error?)
+}
+
 class PaymentMethodFormViewController: UIViewController {
+    /// This caches forms for payment methods so that customers don't have to re-enter details
+    /// This class expects the formCache to be invalidated (cleared) when we load PaymentSheet; we assume the form generated for a given PM type _does not change_ at any point after load.
+    static var formCache: [PaymentSheet.PaymentMethodType: PaymentMethodElement] = [:]
```
Comment (porter-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/ViewControllers/PaymentMethodFormViewController.swift:21]: Struggling to see why this needs to be static

**Review by porter-stripe — APPROVED** [review]
looks good, nothing blocking!

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/New Payment Method Screen/AddPaymentMethodViewController.swift (hunk 1 above)
Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/New Payment Method Screen/AddPaymentMethodViewController.swift:110]: Great q, a lot of code assumed this incorrectly and I just fixed it! We can't; whether link is enabled on the Intent is not the same as whether it's enabled for us. e.g. we disable Link if the merchant requires billing address.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/ViewControllers/PaymentMethodFormViewController.swift (hunk 2 above)
Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/ViewControllers/PaymentMethodFormViewController.swift:21]: We create instances of this class ~every time we go to the PM form, and the cache needs to live longer than that.

**Review by porter-stripe — APPROVED** [review]

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/New Payment Method Screen/AddPaymentMethodViewController.swift (hunk 1 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/New Payment Method Screen/AddPaymentMethodViewController.swift:110]: Makes sense, thanks!

**Review by porter-stripe — APPROVED** [review]

**Pull request merged**